kratos/utilities/math_utils.h

```diff
@@ -976,34 +978,43 @@ class MathUtils
         CrossProduct(c,a,b);
         const double norm = norm_2(c);
 #ifdef KRATOS_DEBUG
-        if(norm < 1000.0*std::numeric_limits<double>::epsilon())
+        if(norm < 1000.0*ZeroTolerance)
             KRATOS_ERROR << "norm is 0 when making the UnitCrossProduct of the vectors " << a << " and " << b << std::endl;
 #endif
         c/=norm;
     }
 
     /**
      * @brief This computes a orthonormal basis from a given vector
+     * @param c The input vector
      * @param a First resulting vector
      * @param b Second resulting vector
-     * @param c The input vector
      * @note Orthonormal basis taken from: http://orbit.dtu.dk/files/126824972/onb_frisvad_jgt2012_v2.pdf
      */
     template< class T1, class T2 , class T3>
     static inline void OrthonormalBasis(const T1& c,T2& a,T3& b ){
         KRATOS_DEBUG_ERROR_IF_NOT(norm_2(c) < (1.0 - 1.0e-6) || norm_2(c) > (1.0 + 1.0e-6)) << "Input should be a normal vector" << std::endl;
-        a[0] = 1.0 - std::pow(c[0], 2)/(1.0 + c[2]);
-        a[1] = - (c[0] * c[1])/(1.0 + c[2]);
-        a[2] = - c[0];
-        const double norm_a = norm_2(a);
-        KRATOS_DEBUG_ERROR_IF_NOT(norm_a < std::numeric_limits<double>::epsilon()) << "Zero norm of the vector" << std::endl;
-        a /= norm_a;
-        b[0] = - (c[0] * c[1])/(1.0 + c[2]);
-        b[1] = 1.0 - std::pow(c[1], 2)/(1.0 + c[2]);
-        b[2] = -c[1];
-        const double norm_b = norm_2(b);
-        KRATOS_DEBUG_ERROR_IF_NOT(norm_b < std::numeric_limits<double>::epsilon()) << "Zero norm of the vector" << std::endl;
-        b /= norm_b;
+        if ((c[2] + 1.0) > ZeroTolerance) {
```

Comment:
std::abs? Is it a problem if c[2] = -0.99999.....?
also, note that comparing abs(a-b) > epsilon is essentially equivalent to a == b, so maybe we need to be a bit more relaxed with the tolerance (up to you, though, I didn't study the complete algorithm)

Reply:
This covers all the cases (this must be activated only for [0,0,-1])